src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -58,35 +86,123 @@ internal sealed class ExternalErrorDiagnosticUpdateSource : IDiagnosticUpdateSou
 
             _diagnosticService = diagnosticService;
 
-            _notificationService = _workspace.Services.GetService<IGlobalOperationNotificationService>();
+            _notificationService = _workspace.Services.GetRequiredService<IGlobalOperationNotificationService>();
         }
 
-        public event EventHandler<BuildProgress> BuildProgressChanged;
-        public event EventHandler<DiagnosticsUpdatedArgs> DiagnosticsUpdated;
+        /// <summary>
+        /// Event generated from the serialized <see cref="_taskQueue"/> whenever the build progress in Visual Studio changes.
+        /// Events are guaranteed to be generated in a serial fashion, but may be invoked on any thread.
+        /// </summary>
+        public event EventHandler<BuildProgress>? BuildProgressChanged;
+
+        /// <summary>
+        /// Event generated from the serialized <see cref="_taskQueue"/> whenever build-only diagnostics are reported during a build in Visual Studio.
+        /// These diagnostics are not supported from intellisense and only get refreshed during actual build.
+        /// </summary>
+        public event EventHandler<DiagnosticsUpdatedArgs>? DiagnosticsUpdated;
+
+        /// <summary>
+        /// Event generated from the serialized <see cref="_taskQueue"/> whenever build-only diagnostics are cleared during a build in Visual Studio.
+        /// These diagnostics are not supported from intellisense and only get refreshed during actual build.
+        /// </summary>
         public event EventHandler DiagnosticsCleared { add { } remove { } }
 
+        /// <summary>
+        /// Indicates if a build is currently in progress inside Visual Studio.
+        /// </summary>
         public bool IsInProgress => BuildInprogressState != null;
```

Comment:
just so i understand, but this is fundamentally a racey API right?  i.e. you could ask this, and get back 'false' but have the actual state at that point be 'true' if we transitioned into building?

Reply:
Yes, that is correct. However, this API is only used from ProjectExternalErrorReporter, which works in sync with `ExternalErrorDiagnosticUpdateSource` during a build. I think it wouldn't be safe for other components to rely on this.